File: src/server/global-assets.ts
import { createHash } from 'node:crypto';

import { encodeNamespace, namespace } from '../core';

import type { RaildropStorage, RaildropStoredObject } from './storage';

export interface GlobalAssetSource {
  alias: string;
  body: Uint8Array;
  filename: string;
  contentType: string;
  namespace: readonly string[];
}

export interface GlobalAssetManifestEntry {
  key: string;
  checksum: string;
  size: number;
  contentType: string;
  filename: string;
}

export interface GlobalAssetManifest {
  version: 1;
  generatedAt: string;
  assets: Record<string, GlobalAssetManifestEntry>;
}

export interface GlobalAssetSyncResult extends GlobalAssetManifest {
  orphanedKeys: string[];
}

export interface GlobalAssetMirrorTarget {
  name: string;
  storage: RaildropStorage;
}

export interface GlobalAssetMirrorResult {
  targets: Record<string, { copied: number; verified: number }>;
  objectCount: number;
  manifestAssetCount: number;
}

export const GLOBAL_MANIFEST_KEY = 'public/global/_raildrop/manifest.json';

const isRecord = (value: unknown): value is Record<string, unknown> =>
  typeof value === 'object' && value !== null && !Array.isArray(value);

export const syncGlobalAssets = async (
  storage: RaildropStorage,
  sources: readonly GlobalAssetSource[]
): Promise<GlobalAssetSyncResult> => {
  const previousManifestObject = await storage.get(GLOBAL_MANIFEST_KEY);
  let previousManifest: unknown = null;
  if (previousManifestObject) {
    try {
      previousManifest = JSON.parse(
        await new Response(previousManifestObject.body as BodyInit).text()
      );
    } catch {
      previousManifest = null;
    }
  }
  const assets: Record<string, GlobalAssetManifestEntry> = {};
  const aliases = new Set<string>();
  for (const source of sources) {
    if (!/^[a-z0-9][a-z0-9-]{0,119}$/.test(source.alias)) {
      throw new Error(`Invalid global asset alias: ${source.alias}`);
    }
    if (aliases.has(source.alias)) throw new Error(`Duplicate global asset alias: ${source.alias}`);
    aliases.add(source.alias);
    namespace(...source.namespace);
    const checksum = createHash('sha256').update(source.body).digest('hex');
    const extension = /\.([a-z0-9]{1,12})$/i.exec(source.filename)?.[1]?.toLowerCase() ?? 'bin';
    const key = `public/global/${encodeNamespace(source.namespace)}/${checksum}/file.${extension}`;
    const existing = await storage.head(key);
    if (!existing) {
      await storage.put({
        key,
        body: source.body,
        contentType: source.contentType,
        cacheControl: 'public, max-age=31536000, immutable',
        metadata: {
          'raildrop-access': 'public',
          'raildrop-retention': 'permanent',
          'raildrop-checksum': checksum,
        },
      });
    }
    const verified = await storage.head(key);
    const verifiedBody = await storage.get(key);
    const verifiedChecksum = verifiedBody
      ? createHash('sha256')
          .update(new Uint8Array(await new Response(verifiedBody.body as BodyInit).arrayBuffer()))
          .digest('hex')
      : null;
    if (
      verified?.size !== source.body.byteLength ||
      verified.contentType !== source.contentType ||
      verified.metadata['raildrop-checksum'] !== checksum ||
      verifiedChecksum !== checksum
    ) {
      throw new Error(`Global asset verification failed for alias ${source.alias}.`);
    }
    assets[source.alias] = {
      key,
      checksum,
      size: source.body.byteLength,
      contentType: source.contentType,
      filename: source.filename,
    };
  }
  const manifest: GlobalAssetManifest = {
    version: 1,
    generatedAt: new Date().toISOString(),
    assets,
  };
  await storage.put({
    key: GLOBAL_MANIFEST_KEY,
    body: JSON.stringify(manifest),
    contentType: 'application/json',
    cacheControl: 'public, max-age=300, stale-while-revalidate=86400',
    metadata: {
      'raildrop-access': 'public',
      'raildrop-retention': 'permanent',
    },
  });
  const activeKeys = new Set(Object.values(assets).map((asset) => asset.key));
  const previousAssets: Record<string, unknown> =
    isRecord(previousManifest) && isRecord(previousManifest.assets) ? previousManifest.assets : {};
  const previousKeys = Object.values(previousAssets).flatMap((asset) =>
    isRecord(asset) && typeof asset.key === 'string' ? [asset.key] : []
  );
  const orphanedKeys = [...new Set(previousKeys)].filter((key) => !activeKeys.has(key));
  return { ...manifest, orphanedKeys };
};

const readObjectBytes = async (storage: RaildropStorage, key: string): Promise<Uint8Array> => {
  const object = await storage.get(key);
  if (!object) throw new Error(`Global asset disappeared while mirroring: ${key}`);
  return new Uint8Array(await new Response(object.body as BodyInit).arrayBuffer());
};

const listAllGlobalObjects = async (storage: RaildropStorage) => {
  const objects = [];
  let continuationToken: string | undefined;
  do {
    const page = await storage.list('public/global/', continuationToken);
    objects.push(...page.objects);
    continuationToken = page.nextToken;
  } while (continuationToken);
  return objects;
};

const parseGlobalManifest = async (
  storage: RaildropStorage
): Promise<GlobalAssetManifest | null> => {
  const object = await storage.get(GLOBAL_MANIFEST_KEY);
  if (!object) return null;
  const parsed: unknown = JSON.parse(await new Response(object.body as BodyInit).text());
  if (!isRecord(parsed) || parsed.version !== 1 || !isRecord(parsed.assets)) {
    throw new Error('Invalid global asset manifest encountered while mirroring.');
  }
  return parsed as unknown as GlobalAssetManifest;
};

export const mirrorGlobalAssets = async (
  targets: readonly GlobalAssetMirrorTarget[]
): Promise<GlobalAssetMirrorResult> => {
  if (targets.length < 2) throw new Error('At least two Raildrop mirror targets are required.');

  const inventories = await Promise.all(
    targets.map(async (target) => ({
      ...target,
      objects: await listAllGlobalObjects(target.storage),
      manifest: await parseGlobalManifest(target.storage),
    }))
  );
  const objectSources = new Map<
    string,
    { storage: RaildropStorage; object: RaildropStoredObject }
  >();
  for (const inventory of inventories) {
    for (const object of inventory.objects) {
      if (object.key === GLOBAL_MANIFEST_KEY) continue;
      const existing = objectSources.get(object.key);
      if (existing && existing.object.size !== object.size) {
        throw new Error(`Global asset conflict for ${object.key}; refusing to overwrite it.`);
      }
      if (!existing) objectSources.set(object.key, { storage: inventory.storage, object });
    }
  }

  const canonicalObjects = new Map<
    string,
    {
      body: Uint8Array;
      checksum: string;
      contentType: string;
      cacheControl?: string;
      metadata: Record<string, string>;
    }
  >();

  await Promise.all(
    [...objectSources.entries()].map(async ([key, source]) => {
      const body = await readObjectBytes(source.storage, key);
      const checksum = createHash('sha256').update(body).digest('hex');
      canonicalObjects.set(key, {
        body,
        checksum,
        contentType: source.object.contentType ?? 'application/octet-stream',
        ...(source.object.cacheControl ? { cacheControl: source.object.cacheControl } : {}),
        metadata: source.object.metadata,
      });
    })
  );

  const assets: Record<string, GlobalAssetManifestEntry> = {};
  for (const inventory of inventories) {
    for (const [alias, entry] of Object.entries(inventory.manifest?.assets ?? {})) {
      const existing = assets[alias];
      if (existing && (existing.key !== entry.key || existing.checksum !== entry.checksum)) {
        throw new Error(`Global asset alias conflict for ${alias}; refusing to merge manifests.`);
      }
      assets[alias] = entry;
    }
  }
  for (const [alias, entry] of Object.entries(assets)) {
    const object = canonicalObjects.get(entry.key);
    if (!object) throw new Error(`Manifest entry ${alias} has no available global object.`);
    if (object.checksum !== entry.checksum || object.body.byteLength !== entry.size) {
      throw new Error(`Manifest entry ${alias} does not match an available global object.`);
    }
  }

  const result: GlobalAssetMirrorResult = {
    targets: {},
    objectCount: canonicalObjects.size + 1,
    manifestAssetCount: Object.keys(assets).length,
  };
  await Promise.all(
    targets.map(async (target) => {
      const outcomes = await Promise.all(
        [...canonicalObjects.entries()].map(async ([key, object]) => {
          const current = await target.storage.get(key);
          const currentBytes = current
            ? new Uint8Array(await new Response(current.body as BodyInit).arrayBuffer())
            : null;
          const currentChecksum = currentBytes
            ? createHash('sha256').update(currentBytes).digest('hex')
            : null;
          if (currentChecksum && currentChecksum !== object.checksum) {
            throw new Error(
              `Global asset conflict for ${key} in ${target.name}; refusing to overwrite it.`
            );
          }
          if (!current) {
            await target.storage.put({
              key,
              body: object.body,
              contentType: object.contentType,
              cacheControl: object.cacheControl,
              metadata: object.metadata,
            });
          }
          const stored = await readObjectBytes(target.storage, key);
          if (createHash('sha256').update(stored).digest('hex') !== object.checksum) {
            throw new Error(`Global asset verification failed for ${key} in ${target.name}.`);
          }
          return { copied: current ? 0 : 1 };
        })
      );
      await target.storage.put({
        key: GLOBAL_MANIFEST_KEY,
        body: JSON.stringify({ version: 1, generatedAt: new Date().toISOString(), assets }),
        contentType: 'application/json',
        cacheControl: 'public, max-age=300, stale-while-revalidate=86400',
        metadata: {
          'raildrop-access': 'public',
          'raildrop-retention': 'permanent',
        },
      });
      result.targets[target.name] = {
        copied: outcomes.reduce((total, outcome) => total + outcome.copied, 0),
        verified: outcomes.length + 1,
      };
    })
  );
  return result;
};

File: sdk/go/globalassets.go
package raildrop

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"regexp"
	"strings"
	"time"
)

type GlobalAssetSource struct {
	Alias       string
	Body        []byte
	Filename    string
	ContentType string
	Namespace   []string
}

type GlobalAssetManifestEntry struct {
	Key         string `json:"key"`
	Checksum    string `json:"checksum"`
	Size        int64  `json:"size"`
	ContentType string `json:"contentType"`
	Filename    string `json:"filename"`
}

type GlobalAssetManifest struct {
	Version     int                                 `json:"version"`
	GeneratedAt string                              `json:"generatedAt"`
	Assets      map[string]GlobalAssetManifestEntry `json:"assets"`
}

type GlobalAssetSyncResult struct {
	GlobalAssetManifest
	OrphanedKeys []string `json:"orphanedKeys"`
}

type GlobalAssetMirrorTarget struct {
	Name    string
	Storage Storage
}

type GlobalAssetMirrorTargetResult struct {
	Copied   int `json:"copied"`
	Verified int `json:"verified"`
}

type GlobalAssetMirrorResult struct {
	Targets            map[string]GlobalAssetMirrorTargetResult `json:"targets"`
	ObjectCount        int                                      `json:"objectCount"`
	ManifestAssetCount int                                      `json:"manifestAssetCount"`
}

var globalAssetAliasPattern = regexp.MustCompile(`^[a-z0-9][a-z0-9-]{0,119}$`)

var sha256Pattern = regexp.MustCompile(`^[a-f0-9]{64}$`)

func readManifestBody(storage Storage, ctx context.Context) ([]byte, bool, error) {
	object, err := storage.Get(ctx, GlobalManifestKey, "")
	if err != nil {
		return nil, false, err
	}
	if object == nil {
		return nil, false, nil
	}
	defer object.Close()
	body, err := io.ReadAll(io.LimitReader(object.Body, 32<<20))
	if err != nil {
		return nil, false, fmt.Errorf("global asset manifest could not be read: %w", err)
	}
	return body, true, nil
}

func parseGlobalManifest(ctx context.Context, storage Storage) (*GlobalAssetManifest, error) {
	body, ok, err := readManifestBody(storage, ctx)
	if err != nil || !ok {
		return nil, err
	}
	var parsed struct {
		Version     *int                                `json:"version"`
		GeneratedAt *string                             `json:"generatedAt"`
		Assets      map[string]GlobalAssetManifestEntry `json:"assets"`
	}
	if err := json.Unmarshal(body, &parsed); err != nil {
		return nil, fmt.Errorf("invalid global asset manifest encountered")
	}
	if parsed.Version == nil || *parsed.Version != 1 || parsed.GeneratedAt == nil || parsed.Assets == nil {
		return nil, fmt.Errorf("invalid global asset manifest encountered")
	}
	for alias, entry := range parsed.Assets {
		if !globalAssetAliasPattern.MatchString(alias) || !isValidManifestEntry(alias, entry) {
			return nil, fmt.Errorf("invalid global asset manifest encountered")
		}
	}
	return &GlobalAssetManifest{
		Version:     *parsed.Version,
		GeneratedAt: *parsed.GeneratedAt,
		Assets:      parsed.Assets,
	}, nil
}

func isValidManifestEntry(alias string, entry GlobalAssetManifestEntry) bool {
	if !strings.HasPrefix(entry.Key, "public/global/") || entry.Key == GlobalManifestKey {
		return false
	}
	if !sha256Pattern.MatchString(entry.Checksum) {
		return false
	}
	if !strings.Contains(entry.Key, "/"+entry.Checksum+"/file.") {
		return false
	}
	if entry.Size < 0 || entry.ContentType == "" || entry.Filename == "" {
		return false
	}
	return true
}

func writeGlobalAssetEntries(ctx context.Context, storage Storage, sources []GlobalAssetSource) (map[string]GlobalAssetManifestEntry, error) {
	assets := make(map[string]GlobalAssetManifestEntry)
	seen := make(map[string]bool)
	for _, source := range sources {
		if !globalAssetAliasPattern.MatchString(source.Alias) {
			return nil, fmt.Errorf("invalid global asset alias: %s", source.Alias)
		}
		if seen[source.Alias] {
			return nil, fmt.Errorf("duplicate global asset alias: %s", source.Alias)
		}
		seen[source.Alias] = true
		if _, err := Namespace(source.Namespace...); err != nil {
			return nil, err
		}
		checksum := sha256Hex(source.Body)
		extension := FileExtensionFor(source.Filename, "")
		namespace, err := EncodeNamespace(source.Namespace...)
		if err != nil {
			return nil, err
		}
		key := "public/global/" + namespace + "/" + checksum + "/file." + extension
		existing, err := storage.Head(ctx, key)
		if err != nil {
			return nil, err
		}
		if existing == nil {
			if _, err := storage.Put(ctx, PutArgs{
				Key:          key,
				Body:         source.Body,
				ContentType:  source.ContentType,
				CacheControl: "public, max-age=31536000, immutable",
				Metadata: map[string]string{
					"raildrop-access":    "public",
					"raildrop-retention": "permanent",
					"raildrop-checksum":  checksum,
				},
			}); err != nil {
				return nil, err
			}
		}
		verified, err := storage.Head(ctx, key)
		if err != nil {
			return nil, err
		}
		verifiedBody, err := storage.Get(ctx, key, "")
		if err != nil {
			return nil, err
		}
		verifiedChecksum := ""
		if verifiedBody != nil {
			verifiedBytes, err := io.ReadAll(verifiedBody.Body)
			closeErr := verifiedBody.Close()
			if err != nil {
				return nil, err
			}
			if closeErr != nil {
				return nil, closeErr
			}
			verifiedChecksum = sha256Hex(verifiedBytes)
		}
		storedChecksum := ""
		if verified != nil {
			storedChecksum = verified.Metadata["raildrop-checksum"]
		}
		if verified == nil ||
			verified.Size != int64(len(source.Body)) ||
			verified.ContentType == nil || *verified.ContentType != source.ContentType ||
			storedChecksum != checksum ||
			verifiedChecksum != checksum {
			return nil, fmt.Errorf("global asset verification failed for alias %s", source.Alias)
		}
		assets[source.Alias] = GlobalAssetManifestEntry{
			Key:         key,
			Checksum:    checksum,
			Size:        int64(len(source.Body)),
			ContentType: source.ContentType,
			Filename:    source.Filename,
		}
	}
	return assets, nil
}

func writeGlobalManifest(ctx context.Context, storage Storage, assets map[string]GlobalAssetManifestEntry) (*GlobalAssetManifest, error) {
	manifest := &GlobalAssetManifest{
		Version:     1,
		GeneratedAt: ISOString(time.Now()),
		Assets:      assets,
	}
	payload, err := MarshalJSON(manifest)
	if err != nil {
		return nil, err
	}
	if _, err := storage.Put(ctx, PutArgs{
		Key:          GlobalManifestKey,
		Body:         payload,
		ContentType:  "application/json",
		CacheControl: "public, max-age=300, stale-while-revalidate=86400",
		Metadata: map[string]string{
			"raildrop-access":    "public",
			"raildrop-retention": "permanent",
		},
	}); err != nil {
		return nil, err
	}
	return manifest, nil
}

func findOrphanedKeys(previous map[string]GlobalAssetManifestEntry, active map[string]GlobalAssetManifestEntry) []string {
	activeKeys := make(map[string]bool)
	for _, asset := range active {
		activeKeys[asset.Key] = true
	}
	previousKeys := make(map[string]bool)
	for _, asset := range previous {
		previousKeys[asset.Key] = true
	}
	var orphaned []string
	for key := range previousKeys {
		if !activeKeys[key] {
			orphaned = append(orphaned, key)
		}
	}
	return orphaned
}

func SyncGlobalAssets(ctx context.Context, storage Storage, sources []GlobalAssetSource) (*GlobalAssetSyncResult, error) {
	previousBody, ok, err := readManifestBody(storage, ctx)
	if err != nil {
		return nil, err
	}
	var previousAssets map[string]GlobalAssetManifestEntry
	if ok {
		var lenient struct {
			Assets map[string]GlobalAssetManifestEntry `json:"assets"`
		}
		if json.Unmarshal(previousBody, &lenient) == nil {
			previousAssets = lenient.Assets
		}
	}
	assets, err := writeGlobalAssetEntries(ctx, storage, sources)
	if err != nil {
		return nil, err
	}
	manifest, err := writeGlobalManifest(ctx, storage, assets)
	if err != nil {
		return nil, err
	}
	if previousAssets == nil {
		previousAssets = map[string]GlobalAssetManifestEntry{}
	}
	return &GlobalAssetSyncResult{GlobalAssetManifest: *manifest, OrphanedKeys: findOrphanedKeys(previousAssets, assets)}, nil
}

func UpsertGlobalAssets(ctx context.Context, storage Storage, sources []GlobalAssetSource) (*GlobalAssetSyncResult, error) {
	previousManifest, err := parseGlobalManifest(ctx, storage)
	if err != nil {
		return nil, err
	}
	upserted, err := writeGlobalAssetEntries(ctx, storage, sources)
	if err != nil {
		return nil, err
	}
	assets := make(map[string]GlobalAssetManifestEntry)
	if previousManifest != nil {
		for alias, entry := range previousManifest.Assets {
			assets[alias] = entry
		}
	}
	for alias, entry := range upserted {
		assets[alias] = entry
	}
	manifest, err := writeGlobalManifest(ctx, storage, assets)
	if err != nil {
		return nil, err
	}
	previous := map[string]GlobalAssetManifestEntry{}
	if previousManifest != nil {
		previous = previousManifest.Assets
	}
	if previous == nil {
		previous = map[string]GlobalAssetManifestEntry{}
	}
	return &GlobalAssetSyncResult{GlobalAssetManifest: *manifest, OrphanedKeys: findOrphanedKeys(previous, assets)}, nil
}

func listAllGlobalObjects(ctx context.Context, storage Storage) ([]StoredObject, error) {
	var objects []StoredObject
	token := ""
	for {
		page, err := storage.List(ctx, "public/global/", token)
		if err != nil {
			return nil, err
		}
		objects = append(objects, page.Objects...)
		if page.NextToken == "" {
			return objects, nil
		}
		token = page.NextToken
	}
}

func readObjectBytes(ctx context.Context, storage Storage, key string) ([]byte, error) {
	object, err := storage.Get(ctx, key, "")
	if err != nil {
		return nil, err
	}
	if object == nil {
		return nil, fmt.Errorf("global asset disappeared while mirroring: %s", key)
	}
	defer object.Close()
	return io.ReadAll(object.Body)
}

func MirrorGlobalAssets(ctx context.Context, targets []GlobalAssetMirrorTarget) (*GlobalAssetMirrorResult, error) {
	if len(targets) < 2 {
		return nil, fmt.Errorf("at least two Raildrop mirror targets are required")
	}
	type inventory struct {
		target   GlobalAssetMirrorTarget
		objects  []StoredObject
		manifest *GlobalAssetManifest
	}
	inventories := make([]inventory, 0, len(targets))
	for _, target := range targets {
		objects, err := listAllGlobalObjects(ctx, target.Storage)
		if err != nil {
			return nil, err
		}
		manifest, err := parseGlobalManifest(ctx, target.Storage)
		if err != nil {
			return nil, err
		}
		inventories = append(inventories, inventory{target: target, objects: objects, manifest: manifest})
	}
	objectSources := make(map[string]GlobalAssetMirrorTarget)
	objectSizes := make(map[string]int64)
	for _, entry := range inventories {
		for _, object := range entry.objects {
			if object.Key == GlobalManifestKey {
				continue
			}
			if _, ok := objectSources[object.Key]; ok {
				if objectSizes[object.Key] != object.Size {
					return nil, fmt.Errorf("global asset conflict for %s; refusing to overwrite it", object.Key)
				}
				continue
			}
			objectSources[object.Key] = entry.target
			objectSizes[object.Key] = object.Size
		}
	}
	type canonicalObject struct {
		body         []byte
		checksum     string
		contentType  string
		cacheControl string
		metadata     map[string]string
	}
	canonicalObjects := make(map[string]canonicalObject)
	for key, source := range objectSources {
		body, err := readObjectBytes(ctx, source.Storage, key)
		if err != nil {
			return nil, err
		}
		stored, err := source.Storage.Head(ctx, key)
		if err != nil {
			return nil, err
		}
		if stored == nil {
			return nil, fmt.Errorf("global asset disappeared while mirroring: %s", key)
		}
		contentType := "application/octet-stream"
		if stored.ContentType != nil {
			contentType = *stored.ContentType
		}
		cacheControl := ""
		if stored.CacheControl != nil {
			cacheControl = *stored.CacheControl
		}
		canonicalObjects[key] = canonicalObject{
			body:         body,
			checksum:     sha256Hex(body),
			contentType:  contentType,
			cacheControl: cacheControl,
			metadata:     stored.Metadata,
		}
	}
	assets := make(map[string]GlobalAssetManifestEntry)
	for _, entry := range inventories {
		if entry.manifest == nil {
			continue
		}
		for alias, manifestEntry := range entry.manifest.Assets {
			if existing, ok := assets[alias]; ok {
				if existing.Key != manifestEntry.Key || existing.Checksum != manifestEntry.Checksum {
					return nil, fmt.Errorf("global asset alias conflict for %s; refusing to merge manifests", alias)
				}
				continue
			}
			assets[alias] = manifestEntry
		}
	}
	for alias, entry := range assets {
		object, ok := canonicalObjects[entry.Key]
		if !ok {
			return nil, fmt.Errorf("manifest entry %s has no available global object", alias)
		}
		if object.checksum != entry.Checksum || int64(len(object.body)) != entry.Size {
			return nil, fmt.Errorf("manifest entry %s does not match an available global object", alias)
		}
	}
	result := &GlobalAssetMirrorResult{
		Targets:            make(map[string]GlobalAssetMirrorTargetResult),
		ObjectCount:        len(canonicalObjects) + 1,
		ManifestAssetCount: len(assets),
	}
	for _, entry := range inventories {
		copied := 0
		for key, object := range canonicalObjects {
			current, err := entry.target.Storage.Get(ctx, key, "")
			if err != nil {
				return nil, err
			}
			currentChecksum := ""
			if current != nil {
				currentBytes, err := io.ReadAll(current.Body)
				_ = current.Close()
				if err != nil {
					return nil, err
				}
				currentChecksum = sha256Hex(currentBytes)
			}
			if currentChecksum != "" && currentChecksum != object.checksum {
				return nil, fmt.Errorf("global asset conflict for %s in %s; refusing to overwrite it", key, entry.target.Name)
			}
			if current == nil {
				putArgs := PutArgs{
					Key:         key,
					Body:        object.body,
					ContentType: object.contentType,
					Metadata:    object.metadata,
				}
				if object.cacheControl != "" {
					putArgs.CacheControl = object.cacheControl
				}
				if _, err := entry.target.Storage.Put(ctx, putArgs); err != nil {
					return nil, err
				}
				copied++
			}
			stored, err := readObjectBytes(ctx, entry.target.Storage, key)
			if err != nil {
				return nil, err
			}
			if sha256Hex(stored) != object.checksum {
				return nil, fmt.Errorf("global asset verification failed for %s in %s", key, entry.target.Name)
			}
		}
		manifest := &GlobalAssetManifest{
			Version:     1,
			GeneratedAt: ISOString(time.Now()),
			Assets:      assets,
		}
		payload, err := MarshalJSON(manifest)
		if err != nil {
			return nil, err
		}
		if _, err := entry.target.Storage.Put(ctx, PutArgs{
			Key:          GlobalManifestKey,
			Body:         payload,
			ContentType:  "application/json",
			CacheControl: "public, max-age=300, stale-while-revalidate=86400",
			Metadata: map[string]string{
				"raildrop-access":    "public",
				"raildrop-retention": "permanent",
			},
		}); err != nil {
			return nil, err
		}
		result.Targets[entry.target.Name] = GlobalAssetMirrorTargetResult{Copied: copied, Verified: len(canonicalObjects) + 1}
	}
	return result, nil
}
